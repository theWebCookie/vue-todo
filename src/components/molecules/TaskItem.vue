<template>
  <li :class="$style.taskItem">
    <InputCheckbox :checked="task.completed" :disabled="false" :id="task.id" />
    <div :class="$style.taskName">
      <label :for="task.id">{{ task.title }}</label>
      <AppButton :customClass="$style.deleteButton" @click="removeTask(task.id)"> X </AppButton>
    </div>
  </li>
</template>

<script>
import AppButton from '../atoms/AppButton.vue';
import InputCheckbox from '../atoms/InputCheckbox.vue';

export default {
  name: 'TaskItem',
  components: {
    InputCheckbox,
    AppButton,
  },
  props: {
    task: {
      type: Object,
      required: true,
    },
  },
  methods: {
    removeTask(taskId) {
      this.$store.dispatch('removeTask', taskId);
    },
  },
};
</script>

<style module lang="scss">
.taskItem {
  display: flex;
  gap: 10px;
  padding: 15px 10px;
  border-bottom: 1px solid #ccc;

  .taskName {
    width: 100%;
    display: flex;
    justify-content: space-between;
    align-items: center;
  }

  &:hover .deleteButton {
    visibility: visible;
  }
}

.deleteButton {
  visibility: hidden;
}
</style>
