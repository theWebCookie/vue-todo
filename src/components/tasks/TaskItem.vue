<template>
  <li :class="$style.taskItem">
    <input type="checkbox" :id="task.id" :checked="task.completed" @change="toggleTask(task.id)" />
    <div :class="$style.taskName">
      <label :for="task.id">{{ task.title }}</label>
      <button :class="$style.deleteButton" @click="removeTask(task.id)">X</button>
    </div>
  </li>
</template>

<script>
export default {
  name: 'TaskItem',
  props: {
    task: {
      type: Object,
      required: true,
    },
  },
  methods: {
    toggleTask(taskId) {
      this.$store.dispatch('toggleTask', taskId);
    },
    removeTask(taskId) {
      this.$store.dispatch('removeTask', taskId);
    },
  },
};
</script>

<style module lang="scss">
.taskItem {
  display: flex;
  gap: 10px;
  padding: 15px 10px;
  border-bottom: 1px solid #ccc;

  .taskName {
    width: 100%;
    display: flex;
    justify-content: space-between;
    align-items: center;

    .deleteButton {
      all: unset;
      cursor: pointer;
      visibility: hidden;
    }
  }

  &:hover .deleteButton {
    visibility: visible;
  }
}
</style>
