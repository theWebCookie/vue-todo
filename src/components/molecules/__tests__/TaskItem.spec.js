import { shallowMount } from '@vue/test-utils';
import { describe, it, expect } from '@jest/globals';
import TaskItem from '../TaskItem.vue';
import { render, fireEvent } from '@testing-library/vue';
import { createStore } from 'vuex';
import { jest } from '@jest/globals';

const styles = {
  taskItem: 'taskItem',
  taskItemText: 'taskItemText',
  taskItemCheckbox: 'taskItemCheckbox',
};

describe('TaskItem.vue', () => {
  it('renders correctly and matches the snapshot', () => {
    const wrapper = shallowMount(TaskItem, {
      global: {
        mocks: {
          $style: styles,
        },
      },
      props: {
        task: {
          id: 1,
          title: 'Task 1',
          completed: false,
        },
      },
    });

    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.classes()).toContain(styles.taskItem);
    const label = wrapper.find('label');
    expect(label.text()).toBe('Task 1');
  });

  it('renders with checked state', () => {
    const wrapper = shallowMount(TaskItem, {
      global: {
        mocks: {
          $style: styles,
        },
      },
      props: {
        task: {
          id: 1,
          title: 'Task 1',
          completed: true,
        },
      },
    });

    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.classes()).toContain(styles.taskItem);
    expect(wrapper.props('task').title).toBe('Task 1');
    expect(wrapper.props('task').completed).toBe(true);
  });

  it('calls removeTask method when delete button is clicked', async () => {
    const task = { id: 1, title: 'Test Task', completed: false };

    let store;
    let actions = {
      removeTask: jest.fn(),
    };

    store = createStore({
      actions,
    });

    const { getByText } = render(TaskItem, {
      global: {
        plugins: [store],
      },
      props: {
        task,
      },
    });

    const deleteButton = getByText('X');
    await fireEvent.click(deleteButton);
    expect(actions.removeTask).toHaveBeenCalledWith(expect.any(Object), 1);
  });
});
