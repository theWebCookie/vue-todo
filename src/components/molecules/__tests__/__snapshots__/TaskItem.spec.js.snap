// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskItem.vue renders correctly and matches the snapshot 1`] = `
<li
  class="taskItem"
>
  <label
    class="taskLabel"
    for="1"
  >
    <input-checkbox-stub
      checked="false"
      disabled="false"
      id="1"
    />
    <div
      class="taskDescription"
    >
      <span
        class="taskTitle"
      >
        Task 1
      </span>
      <app-button-stub
        customclass="deleteButton"
      />
    </div>
  </label>
</li>
`;

exports[`TaskItem.vue renders with checked state 1`] = `
<li
  class="taskItem"
>
  <label
    class="taskLabel"
    for="1"
  >
    <input-checkbox-stub
      checked="true"
      disabled="false"
      id="1"
    />
    <div
      class="taskDescription"
    >
      <span
        class="taskTitle"
      >
        Task 1
      </span>
      <app-button-stub
        customclass="deleteButton"
      />
    </div>
  </label>
</li>
`;
